File: tests/Web.Tests.Playwright/tests/pages/articles-list.page.ts
import { Page, Locator } from '@playwright/test';
import { BasePage } from './base.page';

/**
 * Articles List Page Object Model
 */
export class ArticlesListPage extends BasePage {
  readonly pageHeading: Locator;
  readonly articlesList: Locator;
  readonly articleItems: Locator;
  readonly createButton: Locator;
  readonly searchInput: Locator;
  readonly filterDropdown: Locator;
  readonly noArticlesMessage: Locator;

  constructor(page: Page) {
    super(page);
    this.pageHeading = page.locator('h1, h2').first();
    this.articlesList = page.locator('[data-testid="articles-list"], .articles-list, .article-grid');
    this.articleItems = page.locator('[data-testid="article-item"], .article-item, .article-card');
    this.createButton = page.locator('a[href*="/articles/create"], button:has-text("Create"), button:has-text("New Article")');
    this.searchInput = page.locator('input[type="search"], input[placeholder*="Search"]');
    this.filterDropdown = page.locator('select[name="filter"], select[id*="filter"]');
    this.noArticlesMessage = page.locator('text=/no articles/i, .empty-state');
  }

  /**
   * Navigate to articles list page
   */
  async goto(): Promise<void> {
    await super.goto('/articles');
  }

  /**
   * Get the page heading text
   */
  async getHeadingText(): Promise<string> {
    return await this.pageHeading.textContent() || '';
  }

  /**
   * Get the count of articles displayed
   */
  async getArticlesCount(): Promise<number> {
    return await this.articleItems.count();
  }

  /**
   * Check if articles list is visible
   */
  async hasArticlesList(): Promise<boolean> {
    try {
      return await this.articlesList.isVisible({ timeout: 5000 });
    } catch {
      return false;
    }
  }

  /**
   * Check if create button is visible
   */
  async hasCreateButton(): Promise<boolean> {
    try {
      return await this.createButton.isVisible({ timeout: 5000 });
    } catch {
      return false;
    }
  }

  /**
   * Click on create new article button
   */
  async clickCreateButton(): Promise<void> {
    await this.createButton.first().click();
    await this.page.waitForLoadState('networkidle');
  }

  /**
   * Click on a specific article by index
   */
  async clickArticle(index: number = 0): Promise<void> {
    await this.articleItems.nth(index).click();
    await this.page.waitForLoadState('networkidle');
  }

  /**
   * Search for articles
   */
  async searchArticles(searchTerm: string): Promise<void> {
    if (await this.searchInput.isVisible({ timeout: 2000 })) {
      await this.searchInput.fill(searchTerm);
      await this.page.waitForLoadState('networkidle');
    }
  }

  /**
   * Check if no articles message is displayed
   */
  async hasNoArticlesMessage(): Promise<boolean> {
    try {
      return await this.noArticlesMessage.isVisible({ timeout: 5000 });
    } catch {
      return false;
    }
  }
}

File: e2e/Web.Tests.Playwright/PageObjects/ArticlesListPage.cs
namespace Web.Tests.Playwright.PageObjects;

/// <summary>
/// Articles List Page Object Model
/// </summary>
public class ArticlesListPage : BasePage
{
	private readonly ILocator _pageHeading;
	private readonly ILocator _articlesList;
	private readonly ILocator _articleItems;
	private readonly ILocator _createButton;
	private readonly ILocator _searchInput;
	private readonly ILocator _filterDropdown;
	private readonly ILocator _noArticlesMessage;

	public ArticlesListPage(IPage page) : base(page)
	{
		_pageHeading = page.Locator("h1, h2").First;
		_articlesList = page.Locator("[data-testid='articles-list'], .articles-list, .article-grid");
		_articleItems = page.Locator("[data-testid='article-item'], .article-item, .article-card");
		_createButton = page.Locator("a[href*='/articles/create'], button:has-text('Create'), button:has-text('New Article')");
		_searchInput = page.Locator("input[type='search'], input[placeholder*='Search']");
		_filterDropdown = page.Locator("select[name='filter'], select[id*='filter']");
		_noArticlesMessage = page.Locator("text=/no articles/i, .empty-state");
	}

	/// <summary>
	/// Navigate to articles list page
	/// </summary>
	public override async Task GotoAsync(string path = "/")
	{
		await base.GotoAsync("/articles");
	}

	/// <summary>
	/// Get the page heading text
	/// </summary>
	public async Task<string> GetHeadingTextAsync()
	{
		return await _pageHeading.TextContentAsync() ?? string.Empty;
	}

	/// <summary>
	/// Get the count of articles displayed
	/// </summary>
	public async Task<int> GetArticlesCountAsync()
	{
		return await _articleItems.CountAsync();
	}

	/// <summary>
	/// Check if articles list is visible
	/// </summary>
	[Obsolete]
	public async Task<bool> HasArticlesListAsync()
	{
		try
		{
			return await _articlesList.IsVisibleAsync(new LocatorIsVisibleOptions { Timeout = 5000 });
		}
		catch
		{
			return false;
		}
	}

	/// <summary>
	/// Check if create button is visible
	/// </summary>
	[Obsolete]
	public async Task<bool> HasCreateButtonAsync()
	{
		try
		{
			LocatorIsVisibleOptions options = new LocatorIsVisibleOptions() ;
			options.Timeout = 5000;

			return await _createButton.IsVisibleAsync(options);
		}
		catch
		{
			return false;
		}
	}

	/// <summary>
	/// Click on create new article button
	/// </summary>
	public async Task ClickCreateButtonAsync()
	{
		await _createButton.First.ClickAsync();
		await Page.WaitForLoadStateAsync(LoadState.NetworkIdle);
	}

	/// <summary>
	/// Click on a specific article by index
	/// </summary>
	public async Task ClickArticleAsync(int index = 0)
	{
		await _articleItems.Nth(index).ClickAsync();
		await Page.WaitForLoadStateAsync(LoadState.NetworkIdle);
	}

	/// <summary>
	/// Search for articles
	/// </summary>
	[Obsolete]
	public async Task SearchArticlesAsync(string searchTerm)
	{
		if (await _searchInput.IsVisibleAsync(new LocatorIsVisibleOptions { Timeout = 2000 }))
		{
			await _searchInput.FillAsync(searchTerm);
			await Page.WaitForLoadStateAsync(LoadState.NetworkIdle);
		}
	}

	/// <summary>
	/// Check if no articles message is displayed
	/// </summary>
	[Obsolete]
	public async Task<bool> HasNoArticlesMessageAsync()
	{
		try
		{
			return await _noArticlesMessage.IsVisibleAsync(new LocatorIsVisibleOptions { Timeout = 5000 });
		}
		catch
		{
			return false;
		}
	}

	// Add this public property to expose _searchInput safely
	public ILocator? SearchInput => _searchInput;
}